relation_engine_server/test/test_api_v1.py

```diff
@@ -104,57 +136,248 @@ def test_update_specs(self):
+        # under the key 'stored_query'
+        def check_resp_json(self, resp):
+            resp_json = resp.json()
+            self.assertEqual(type(resp_json['stored_query']), dict)
+            self.assertEqual(resp_json['stored_query']['name'], name)
+            self.assertEqual(set(resp_json['stored_query'].keys()), {
+                'name', 'query', 'params'
+            })
+
+        self.test_request(
+            '/specs/stored_queries',
+            {'name': name},
+            resp_test=check_resp_json
+        )
+
+    def test_fetch_invalid_data_source(self):
+        """Unknown data source name should yield 404 status."""
 
-    def test_fetch_invalid_queries(self):
+        name = 'invalid_data_source'
+        self.test_request(
+            '/specs/data_sources',
+            {'name': name},
+            status_code=404,
+            resp_json={
+                'error': {
+                    'status': 404,
+                    'message': 'Not found',
+                    'details': f"Data source '{name}' does not exist.",
+                    'name': name,
+                }
+            }
+        )
+
+    def test_fetch_invalid_collections_and_documents(self):
+        """Test the case where the collection or document does not exist."""
+
+        name = 'fake_collection'
+        collection_params = {'name': name}  # fetch an invalid collection
+        document_params = {'doc_id': name + '/123'}  # fetch an invalid document
+        for variant in ['schemas', 'collections']:
+            for params in [document_params, collection_params]:
+
+                self.test_request(
+                    '/specs/' + variant,
+                    params=params,
+                    status_code=404,
+                    resp_json={
+                        'error': {
+                            'status': 404,
+                            'message': 'Not found',
+                            'details': f"Collection '{name}' does not exist.",
+                            'name': name,
+                        }
+                    }
+                )
+
+    def test_fetch_invalid_stored_queries(self):
         """Test the case where the stored query does not exist."""
-        resp = requests.get(
-            API_URL + '/specs/stored_queries',
-            params={'name': 'xyzabc'},
-        ).json()
-        self.assertEqual(resp['error'], 'Stored query does not exist.')
+
+        name = 'made_up_stored_query'
+        self.test_request(
+            '/specs/stored_queries',
+            params={'name': name},
+            status_code=404,
+            resp_json={
+                'error': {
+                    'status': 404,
+                    'message': 'Not found',
+                    'details': f"Stored query '{name}' does not exist.",
+                    'name': name,
+                }
+            }
+        )
+
+    def test_show_data_sources(self):
```

Comment:
these next three tests are for the pre-existing endpoints; I used the test helper for two of the tests, but what is being done in the tests has not changed.